osu.Server.Queues.ScoreStatisticsProcessor.Tests/DatabaseTest.cs

```diff
@@ -170,6 +172,43 @@ protected Beatmap AddBeatmap(Action<Beatmap>? beatmapSetup = null, Action<Beatma
             return beatmap;
         }
 
+        protected void AddBeatmapAttributes(Action<DifficultyAttributes>? setup = null)

```

Comment:
Maybe this is gonna be nitpicking but for a helper extracted this high it doesn't seem to be terribly usable in conjunction with everything else.

- (Most relevant) Hardcodes `TEST_BEATMAP_ID` even though other tests allow the id to be changed already
- (Less relevant) Hardcodes osu! difficulty attributes

Honestly I'm not sure the method needs to live that high up, I'd move it back to `PerformanceProcessorTests` as a local helper until actually needed elsewhere. That said If I were to suggest how I'd see this as a more general helper, it'd probably be something along the lines of

```diff
diff --git a/osu.Server.Queues.ScoreStatisticsProcessor.Tests/DatabaseTest.cs b/osu.Server.Queues.ScoreStatisticsProcessor.Tests/DatabaseTest.cs
index aaf29b5..7be5a50 100644
--- a/osu.Server.Queues.ScoreStatisticsProcessor.Tests/DatabaseTest.cs
+++ b/osu.Server.Queues.ScoreStatisticsProcessor.Tests/DatabaseTest.cs
@@ -12,7 +12,6 @@ using osu.Framework.Extensions.ExceptionExtensions;
 using osu.Game.Beatmaps;
 using osu.Game.Online.API.Requests.Responses;
 using osu.Game.Rulesets.Difficulty;
-using osu.Game.Rulesets.Osu.Difficulty;
 using osu.Game.Rulesets.Scoring;
 using osu.Game.Scoring;
 using osu.Server.Queues.ScoreStatisticsProcessor.Models;
@@ -172,23 +171,13 @@ namespace osu.Server.Queues.ScoreStatisticsProcessor.Tests
             return beatmap;
         }
 
-        protected void AddBeatmapAttributes(Action<DifficultyAttributes>? setup = null)
+        protected void AddBeatmapAttributes<TDifficultyAttributes>(uint? beatmapId = null, Action<TDifficultyAttributes>? setup = null)
+            where TDifficultyAttributes : DifficultyAttributes, new()
         {
-            var attribs = new OsuDifficultyAttributes
+            var attribs = new TDifficultyAttributes
             {
                 StarRating = 5,
                 MaxCombo = 5,
-                AimDifficulty = 5,
-                SpeedDifficulty = 5,
-                SpeedNoteCount = 5,
-                FlashlightDifficulty = 5,
-                SliderFactor = 5,
-                ApproachRate = 5,
-                OverallDifficulty = 5,
-                DrainRate = 5,
-                HitCircleCount = 5,
-                SliderCount = 5,
-                SpinnerCount = 5
             };
 
             setup?.Invoke(attribs);
@@ -199,7 +188,7 @@ namespace osu.Server.Queues.ScoreStatisticsProcessor.Tests
                 {
                     db.Insert(new BeatmapDifficultyAttribute
                     {
-                        beatmap_id = TEST_BEATMAP_ID,
+                        beatmap_id = beatmapId ?? TEST_BEATMAP_ID,
                         mode = 0,
                         mods = 0,
                         attrib_id = (ushort)a.attributeId,
diff --git a/osu.Server.Queues.ScoreStatisticsProcessor.Tests/PerformanceProcessorTests.cs b/osu.Server.Queues.ScoreStatisticsProcessor.Tests/PerformanceProcessorTests.cs
index 00eff66..ba0b66b 100644
--- a/osu.Server.Queues.ScoreStatisticsProcessor.Tests/PerformanceProcessorTests.cs
+++ b/osu.Server.Queues.ScoreStatisticsProcessor.Tests/PerformanceProcessorTests.cs
@@ -6,6 +6,7 @@ using osu.Framework.Extensions.TypeExtensions;
 using osu.Game.Rulesets.Catch.Mods;
 using osu.Game.Rulesets.Mania.Mods;
 using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Osu.Difficulty;
 using osu.Game.Rulesets.Osu.Mods;
 using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.Taiko.Mods;
@@ -29,7 +30,7 @@ namespace osu.Server.Queues.ScoreStatisticsProcessor.Tests
         public void PerformanceIndexUpdates()
         {
             AddBeatmap();
-            AddBeatmapAttributes();
+            AddBeatmapAttributes<OsuDifficultyAttributes>();
 
             SetScoreForBeatmap(TEST_BEATMAP_ID, score =>
             {
@@ -192,7 +193,7 @@ namespace osu.Server.Queues.ScoreStatisticsProcessor.Tests
         public void FailedScoreDoesNotProcess()
         {
             AddBeatmap();
-            AddBeatmapAttributes();
+            AddBeatmapAttributes<OsuDifficultyAttributes>();
 
             ScoreItem score = SetScoreForBeatmap(TEST_BEATMAP_ID, score =>
             {

```

Reply:
I thought there was a reason why I specified all the other values - perhaps they're required for PP to be non-zero, But it's cool that this works as is.